File: app/src/main/java/com/amaze/filemanager/services/ssh/SFtpClientTemplate.java
package com.amaze.filemanager.services.ssh;

import android.support.annotation.NonNull;

import net.schmizz.sshj.sftp.SFTPClient;

import java.io.IOException;

public abstract class SFtpClientTemplate<T>
{
    public final String url;

    public final boolean closeClientOnFinish;

    public SFtpClientTemplate(@NonNull String url)
    {
        this(url, true);
    }

    public SFtpClientTemplate(@NonNull String url, boolean closeClientOnFinish)
    {
        this.url = url;
        this.closeClientOnFinish = closeClientOnFinish;
    }

    public abstract T execute(@NonNull SFTPClient client) throws IOException;
}

File: file_operations/src/main/java/com/amaze/filemanager/file_operations/filesystem/filetypes/sftp/templates/SFtpClientTemplate.kt
package com.amaze.filemanager.file_operations.filesystem.filetypes.sftp.templates

import net.schmizz.sshj.sftp.SFTPClient
import java.io.IOException

/**
 * Template class for executing actions with [SFTPClient] while leave the complexities of
 * handling connection and session setup/teardown to [SshClientUtils].
 */
abstract class SFtpClientTemplate <T>
/**
 * If closeClientOnFinish is set to true, calling code needs to handle closing of [ ] session.
 *
 * @param url SSH connection URL, in the form of `
 * ssh://<username>:<password>@<host>:<port>` or `
 * ssh://<username>@<host>:<port>`
 */ @JvmOverloads
constructor(@JvmField val iv: String, @JvmField val url: String, @JvmField val closeClientOnFinish: Boolean = true) {

    /**
     * Implement logic here.
     *
     * @param client [SFTPClient] instance, with connection opened and authenticated, and SSH
     * session had been set up.
     * @param <T> Requested return type
     * @return Result of the execution of the type requested
     */
    @Throws(IOException::class)
    abstract fun execute(client: SFTPClient): T
}